baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/CardinalityEqualsZero.java

```diff
@@ -0,0 +1,140 @@
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.matchers.Matcher;
+import com.google.errorprone.matchers.Matchers;
+import com.google.errorprone.matchers.method.MethodMatchers;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.BinaryTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.IdentifierTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.tools.javac.tree.JCTree;
+import java.util.Collection;
+import java.util.Optional;
+import org.immutables.value.Value.Immutable;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        severity = SeverityLevel.SUGGESTION,
+        summary = "Use the isEmpty method instead of checking collection size")
+public final class CardinalityEqualsZero extends BugChecker implements BugChecker.BinaryTreeMatcher {
+    private static final Matcher<ExpressionTree> COLLECTION_SIZE_METHOD_MATCHER = MethodMatchers.instanceMethod()
+            .onDescendantOf(Collection.class.getName())
+            .named("size")
+            .withNoParameters();
+
+    private static final Matcher<ExpressionTree> INT_ZERO = Matchers.intLiteral(0);
+
+    @Override
+    public Description matchBinary(BinaryTree tree, VisitorState state) {
+        Optional<EqualsZeroExpression> maybeEqualsZeroExpression = getEqualsZeroExpression(tree, state);
+        if (maybeEqualsZeroExpression.isEmpty()) {
+            return Description.NO_MATCH;
+        }
+
+        EqualsZeroExpression equalsZeroExpression = maybeEqualsZeroExpression.get();
+        ExpressionTree operand = equalsZeroExpression.operand();
+        ExpressionTree collectionInstance = ASTHelpers.getReceiver(operand);
+
+        if (isInIsEmptyDefinition(state) && (collectionInstance == null || isExpressionThis(collectionInstance))) {
+            return Description.NO_MATCH;
+        }
+
+        if (COLLECTION_SIZE_METHOD_MATCHER.matches(operand, state)) {
+            return describeMatch(
+                    tree,
+                    SuggestedFix.replace(
+                            tree,
+                            (equalsZeroExpression.type() == ExpressionType.NEQ ? "!" : "")
+                                    + (collectionInstance != null
+                                            ? state.getSourceForNode(collectionInstance) + "."
+                                            : "")
+                                    + "isEmpty()"));
+        }
+
+        return Description.NO_MATCH;
+    }
+
+    private static boolean isInIsEmptyDefinition(VisitorState state) {
+        MethodTree methodTree = ASTHelpers.findEnclosingNode(state.getPath(), MethodTree.class);
+        return methodTree != null && methodTree.getName().contentEquals("isEmpty");
+    }
```

Comment:
I'd probably not include this in the condition due to potential edge cases (method delegation to private functions, etc).

Reply:
I'd rather allow potential failures through than produce false positives (which erode trust in the check).

Reply:
I've removed this check. Now, the check does not match when the expression looks like `this.size() == 0` or `size() == 0`, regardless of the method.